### ISSUE TITLE
Erorr when stopping debugging

### ISSUE THREAD

**Issue by Vaccano** [microsoft/mindaro#201]
**Describe the bug**
 
**To Reproduce**
1. Connect my Web API ASP.Net Core 3.1 service to my Kubernetes Cluster via Bridge to Kubernetes
2. Start debugging.
3. Debug for a while
4. Press the stop button.
5. A few seconds after the stop button is pressed, an banner like this appears in Visual Studio:
![image](https://user-images.githubusercontent.com/206109/125132950-53415580-e0c2-11eb-9611-ed9e186d0e4e.png)
(It says "Visual Studio ran into an unexpected problem, and may affect some project system functions." An it has a "Open log file" button.)

When I open the log file, this is what is in it: 

<details>
  <summary>Click to Expand Log File Details</summary>
===================
7/9/2021 2:26:15 PM
Recoverable
System.AggregateException: One or more errors occurred. ---> StreamJsonRpc.RemoteInvocationException: Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. System.InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.ReconnectingConnectionState.WaitForActiveConnectionAsync(String methodName, String memberName, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass31_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
System.InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.ReconnectingConnectionState.WaitForActiveConnectionAsync(String methodName, String memberName, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass31_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ConnectionManager.ConnectionManagerService.<DisconnectCoreAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ConnectionManager.ConnectionManagerService.<>c__DisplayClass27_0.<<GetDisconnectCallback>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ConnectionManager.ConnectionManagerService.<>c__DisplayClass27_0.<<GetDisconnectCallback>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ServiceHub.ConnectServiceBrokerClient.<ExecuteWithServiceAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception ex)
   at Microsoft.VisualStudio.ProjectSystem.ProjectErrorReporting.<>c__DisplayClass6_0.<SubmitErrorReport>b__0()
   at Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__Action|0()
   at GuardMethodClass.GuardMethod(Func`1 , Func`2 , Func`2 )
   --- End of inner exception stack trace ---
---> (Inner Exception #0) StreamJsonRpc.RemoteInvocationException: Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. System.InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.ReconnectingConnectionState.WaitForActiveConnectionAsync(String methodName, String memberName, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass31_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
System.InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.ReconnectingConnectionState.WaitForActiveConnectionAsync(String methodName, String memberName, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass31_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ConnectionManager.ConnectionManagerService.<DisconnectCoreAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ConnectionManager.ConnectionManagerService.<>c__DisplayClass27_0.<<GetDisconnectCallback>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ConnectionManager.ConnectionManagerService.<>c__DisplayClass27_0.<<GetDisconnectCallback>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Kubernetes.Debugging.Core.Services.ServiceHub.ConnectServiceBrokerClient.<ExecuteWithServiceAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception ex)
   at Microsoft.VisualStudio.ProjectSystem.ProjectErrorReporting.<>c__DisplayClass6_0.<SubmitErrorReport>b__0()
   at Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__Action|0()
   at GuardMethodClass.GuardMethod(Func`1 , Func`2 , Func`2 )
RPC server exception:
Microsoft.VisualStudio.Kubernetes.Debugging.RpcContracts.DscException: Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. System.InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.ReconnectingConnectionState.WaitForActiveConnectionAsync(String methodName, String memberName, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass31_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
System.InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.ReconnectingConnectionState.WaitForActiveConnectionAsync(String methodName, String memberName, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass31_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

      at Microsoft.VisualStudio.Kubernetes.Debugging.ServiceHub.DscClient.<InvokeSilentAsync>d__16.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.VisualStudio.Kubernetes.Debugging.ServiceHub.DscClient.<ConnectAsync>d__18.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.VisualStudio.Kubernetes.Debugging.ServiceHub.Connect.ConnectService.<>c__DisplayClass11_0.<<DisconnectAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.VisualStudio.Kubernetes.Debugging.RpcContracts.Telemetry.KubeConnectScopeBase`1.<ExecuteWithErrorHandlingAsync>d__26.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at Microsoft.VisualStudio.Kubernetes.Debugging.RpcContracts.Telemetry.KubeConnectScopeBase`1.<ExecuteWithErrorHandlingAsync>d__26.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.VisualStudio.Kubernetes.Debugging.ServiceHub.Connect.ConnectService.<DisconnectAsync>d__11.MoveNext()
<---

===================
</details>

 
**Expected behavior**
This does not seem to cause any issues, so the only expected behavior is that the error banner would not show.
 
**Logs**
Logs had my company info in them.  I would be happy to provide logs privately if desired.
 
**Environment Details**
Client used: **Visual Studio**
Visual Studio version: **Version 16.9.2**
Client's version: Windows 10 Version - 10.0.19042 Build 19042
Operating System: Windows 10 Version - 10.0.19042 Build 19042
 
**Additional context**
No additional context

**Comment by daniv-msft** [comment]
Thanks @Vaccano for logging this issue!
This `InvalidOperationException: The 'InvokeCoreAsync' method cannot be called if the connection is not active` is an error we've seen in our telemetry but that we could hardly reproduce. Adding @lolodi for FYI, as he's looking at this bug.
One of our hypotheses is that it was related to very long connections (let's say staying connected using Bridge for more than an hour). @Vaccano Could you please confirm if it is the case for you? Or does it reproduce as well for small ones?

Regarding the banner you see/the fact that you cannot debug again right after, that's something we should improve indeed.
@danegsta Could you please let us know if you have an idea or where these might come from? Do we handle errors differently when stopping the connection (shouldn't we swallow this error and just log it on our side)?

**Comment by Vaccano** [comment]
It reproduces with much shorter debugging times.  I just had it happen with a less than 5 minute debugging session.

**Comment by daniv-msft** [comment]
OK thanks, it's definitely good to know! @lolodi on our side is looking into it, and we might ask you for more information if required.

**Comment by Rob623** [comment]
I`m also getting the same error and after it happens VS has to be restarted because it... not working right anymore... text is missing where hints should be - getting empty spaces in text editor... and most importantly, its getting unresponsive to an unusable level. 
I`m guessing that the next VS extension is being run on the same thread and once BTK throw the exception, the extensions loop is broken or something similar. 

@Vaccano did you see such behavior as well? if not, are you using additional extension? I suspect that resharper is the one getting broken in my case. The reason is that it influences the text editor in such a way that can result in the unusable state I`m getting.

**Comment by Vaccano** [comment]
@Rob623 - I have not seen any such issues.  I do use ReSharper, but it has not interacted with Bridge to Kubernetes in any way I can see.  Nor have I seen any issues with with my text editing in Visual Studio.

As I said in my bug report, starting up debugging right after stopping does not work out well.  I got issues where my Kubernetes application service is put in an unstable state.  I got error messages complaining about replicas and I had to re-deploy the application to Kubernetes.  But these symptoms are not as what you are reporting.  (And this is easily avoided by waiting a minute or two before starting to debug again.)

**Comment by Rob623** [comment]
Hmm I see. since the stack trace looks the same, I`m guessing that the issues are related however, since the symptoms appear to be different, @daniv-msft should I open a separate issue for it? 
btw I forgot to mention that I`ve also seen the problem @Vaccano mentioned. 
Anyway, if I find more info or better reproduction scenario I`ll update.

**Comment by daniv-msft** [comment]
@Rob623 Thanks, I think it's fine to keep the same issue for now as even though the symptoms are slightly different, I believe they have the same root cause.
@lolodi made good progress on a tentative fix for this, and we will work to integrate them in Visual Studio to get the changes to you.

**Comment by tigerquoll** [comment]
I've seeing the same issues here as well.
[mindaro-vscode-2021-07-14T05-47-42.911Z.txt](https://github.com/microsoft/mindaro/files/6813854/mindaro-vscode-2021-07-14T05-47-42.911Z.txt)

**Comment by tigerquoll** [comment]
@daniv-msft @lolodi is there any update on the issue?

**Comment by daniv-msft** [comment]
@tigerquoll Thanks for following up! We're releasing our new version (including the fix from @lolodi) in VS Code this week. It's already available to 10% of the customers, and we will gradually increase this if we don't detect issues.
Visual Studio will come right after, so likely in 1/1.5 weeks.
